### PR TITLE
Implement FILE_MIGRATE handling

### PR DESCRIPTION
**Pull request by guzkodes**
This PR implements handling of ```FILE_MIGRATE``` errors when trying to download a file that lives in a different datacenter:

  * Maintains a ```HashMap<i32, FileDownloader>``` at client level
  * Detects the error, picks the ```Sender``` from the session datacenter, exports the auth and imports into a new ```Sender``` that will be stored in the ```HashMap```
  * Redo the request
  
  
  ~~WIP: Depends on https://github.com/Lonami/grammers/pull/146~~

### CODE REVIEW

**Comment by guzkodes** [comment]
The CI error is in a file I din't change, so likely unrelated to the changes

File: lib/grammers-client/examples/downloader.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+//! cargo run --example downloader -- CHAT_NAME
+//! ```
+//!
+//! Messages will be printed to stdout, and media will be saved in the ```target/``` folder locally, named
+//! message-[id].[jpg/mp4]
+//!
+extern crate core;
+extern crate grammers_client;
+extern crate grammers_session;
+extern crate grammers_tl_types;
```
Comment (Lonami) [inline lib/grammers-client/examples/downloader.rs:19]: Is there any reason to do this? None of the other crate examples do.

File: lib/grammers-client/examples/downloader.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    let mut client_handle = client.clone();
+    let network_handle = task::spawn(async move { client.run_until_disconnected().await });
+
+    let chat = client_handle.resolve_username(chat_name.as_str()).await?;
+
+    if chat.is_none() {
+        panic!("Chat {} could not be found", chat_name);
+    }
+
+    let mut messages = client_handle.iter_messages(&chat.unwrap());
```
Comment (Lonami) [inline lib/grammers-client/examples/downloader.rs:112]: `is_none` followed by `unwrap` is not very idiomatic. We could change it with `.expect("Chat could not be found")` or maybe a `match`.

File: lib/grammers-client/examples/downloader.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    return format!(".{}", suffix);
+}
+
+fn get_subtype(mime_type: Option<&str>) -> String {
+    return mime_type
+        .map(|m| {
+            let mime: Mime = m.parse().unwrap();
+            mime.subtype().to_string()
+        })
+        .unwrap_or(String::new());
```
Comment (Lonami) [inline lib/grammers-client/examples/downloader.rs:180]: This might create files such as `message-1001.` (empty extension). Perhaps it would be better to use the extension from the media's file name?

File: lib/grammers-client/src/client/client.rs [hunk 4]
```diff
@@ -119,6 +119,14 @@ pub(crate) struct ClientInner {
     pub(crate) updates: Mutex<VecDeque<crate::types::Update>>,
     // Used to avoid locking the entire sender when enqueueing requests.
     pub(crate) request_tx: Mutex<Enqueuer>,
+    // Stores per-datacenter FileDownloader instances
+    pub(crate) downloader_map: RwLock<HashMap<i32, Arc<FileDownloader>>>,
+}
+
+pub(crate) struct FileDownloader {
```
Comment (Lonami) [inline lib/grammers-client/src/client/client.rs:126]: Not entirely sure if we should name these `FileDownloader`. They're more like generic exported senders after all (which are used for downloads, yes).

File: lib/grammers-client/src/client/files.rs [hunk 5]
```diff
@@ -264,14 +264,10 @@ impl Client {
                             );
                         }
                         Err(InvocationError::Rpc(err)) => {
-                            // Retry on rate limit
-                            if err.code == 420 {
```
Comment (Lonami) [inline lib/grammers-client/src/client/files.rs:268]: Was it intended to remove this retry logic in this PR?

File: lib/grammers-client/src/client/net.rs [hunk 6]
```diff
@@ -263,20 +284,85 @@ impl Client {
...
+                    &new_downloader.sender,
+                    &new_downloader.stepping_done,
+                )
+                .await?;
+
+                mutex.insert(dc_id, new_downloader.clone());
+                Ok(new_downloader.clone())
+            }
+            Err(e) => {
+                panic!("Cannot obtain new sender to datacenter {}", e)
```
Comment (Lonami) [inline lib/grammers-client/src/client/net.rs:332]: Do we really want a panic here, or should this return the error?

File: lib/grammers-client/src/types/media.rs [hunk 7]
```diff
@@ -49,6 +49,94 @@ pub enum Media {
     Contact(Contact),
 }
 
+pub trait MultiSized {
```
Comment (Lonami) [inline lib/grammers-client/src/types/media.rs:52]: I'm okay with improving the way photo sizes but I think this should be in a separate PR (perhaps merged first).

**Review by Lonami — COMMENTED** [review]
Thanks for the PR! Left some comments. My main concern is with some changes that I would prefer for them to be in a separate PR but I'm not sure if there's a reason why they're in this one (single it's also a single commit, it's harder to justify), along with a questionable public API surface.

File: lib/grammers-client/examples/downloader.rs (hunk 1 above)
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:19]: no reason. It is a leftover

File: lib/grammers-client/examples/downloader.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                    );
+                    let path = Path::new(dest.as_str());
+                    client_handle
+                        .download_media(&media, &path)
+                        .await
+                        .expect("Error downloading message");
+                }
+            }
+            Err(e) => panic!("{}", e),
+            Ok(None) => break { println!("Finished iterating messages") },
```
[the remark this answers is not in the document]
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:140]: It does work, but I did it initially like this for debugging purposes, I will refactor to make it more easy on the eyes

File: lib/grammers-client/examples/downloader.rs (hunk 3 above)
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:180]: wdym, it will always try to use the mime type if it's present, otherwise it is not possible to know the extension unless inspecting the bytes of the content

File: lib/grammers-client/src/client/auth.rs [hunk 9]
```diff
@@ -180,6 +181,24 @@ impl Client {
         }
     }
 
+    /// Returns data for copying authorization to another data-center.
+    ///
+    /// See https://core.telegram.org/method/auth.exportAuthorization
+    pub async fn export_authorization(
```
[the remark this answers is not in the document]
Comment (guzkodes) [reply lib/grammers-client/src/client/auth.rs:187]: Initially it was ```pub(crate)``` , I made it ```pub``` because it is part of the telegram API schema (https://core.telegram.org/method/auth.exportAuthorization).
Since the import is non trivial and depends on lots of internal private parts, you're right, I see no reason to expose this.
I will hide this inside ```net.rs```

File: lib/grammers-client/src/client/client.rs (hunk 4 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/client.rs:126]: what about ```DatacenterConnector``` ?

File: lib/grammers-client/src/client/files.rs (hunk 5 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/files.rs:268]: yes, the wait login is inside https://github.com/Lonami/grammers/blob/master/lib/grammers-client/src/client/net.rs#L251-L267

File: lib/grammers-client/src/client/net.rs (hunk 6 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/net.rs:332]: No reason for panic 👍

File: lib/grammers-client/src/types/media.rs (hunk 7 above)
Comment (guzkodes) [reply lib/grammers-client/src/types/media.rs:52]: no problem. I put it all together because without it most downloads fail

File: lib/grammers-client/src/client/net.rs (hunk 6 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/net.rs:332]: How should this ```AuthorizationError``` should be propagated ? it is not compatible with ```InvocationError```

Comment (guzkodes) [reply lib/grammers-client/src/client/net.rs:332]: I am thinking in adding a new variant to ```InvocationError``` to signal for security issues

File: lib/grammers-client/src/types/media.rs (hunk 7 above)
Comment (guzkodes) [reply lib/grammers-client/src/types/media.rs:52]: https://github.com/Lonami/grammers/pull/146

File: lib/grammers-client/examples/downloader.rs (hunk 3 above)
Comment (Lonami) [reply lib/grammers-client/examples/downloader.rs:180]: I meant there would be a trailing dot (`message-1001.` rather than `message-1001`).

File: lib/grammers-client/src/client/files.rs (hunk 5 above)
Comment (Lonami) [reply lib/grammers-client/src/client/files.rs:268]: Yes but this change is still not directly related to handling `FILE_MIGRATE` which is what I meant (if this change was for this or another PR).

File: lib/grammers-client/src/client/client.rs (hunk 4 above)
Comment (Lonami) [reply lib/grammers-client/src/client/client.rs:126]: Hmm not really... Perhaps we can leave it for now.

File: lib/grammers-client/src/client/net.rs (hunk 6 above)
Comment (Lonami) [reply lib/grammers-client/src/client/net.rs:332]: The method should return `AuthorizationError` instead (because `InvocationError` is compatible with it, just not the other way around).

File: lib/grammers-client/examples/downloader.rs (hunk 3 above)
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:180]: Fixed in https://github.com/Lonami/grammers/pull/148

File: lib/grammers-client/examples/downloader.rs [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+//! [Windows](https://ss64.com/nt/set.html) or [Linux](https://ss64.com/bash/export.html))
+//! to Telegram's API ID and API hash respectively.
+//!
+//! Then, run it as:
+//!
+//! ```sh
+//! cargo run --example downloader -- CHAT_NAME
+//! ```
+//!
+//! Messages will be printed to stdout, and media will be saved in the ```target/``` folder locally, named
```
[the remark this answers is not in the document]
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:13]: This was already solved in a separate PR

File: lib/grammers-client/examples/downloader.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (guzkodes) [reply lib/grammers-client/examples/downloader.rs:140]: Already solved on separated  PR

File: lib/grammers-client/src/client/files.rs (hunk 5 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/files.rs:268]: I separated it into another PR: https://github.com/Lonami/grammers/pull/149

File: lib/grammers-client/src/client/net.rs (hunk 6 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/net.rs:332]: My question was more like in terms of API breakage,  as this is method is part a chain of methods:
download_media (file.rs)-> next (file.rs) -> invoke_in_dc (net.rs) -> connect_sender (net.rs)
If I change it to return ```AuthorizationError``` the callers will also have to return it and so on

File: lib/grammers-client/src/client/files.rs [hunk 11]
```diff
@@ -21,6 +21,7 @@ use tokio::{
 
 pub const MIN_CHUNK_SIZE: i32 = 4 * 1024;
 pub const MAX_CHUNK_SIZE: i32 = 512 * 1024;
+pub const FILE_MIGRATE_ERROR: i32 = 303;
```
Comment (Lonami) [inline lib/grammers-client/src/client/files.rs:24]: The chunk size is public to let the users know that's the limits they can configure but this error doesn't really need to be exposed to the users:

```suggestion
const FILE_MIGRATE_ERROR: i32 = 303;
```

File: lib/grammers-client/src/client/net.rs [hunk 12]
```diff
@@ -278,6 +299,84 @@ impl Client {
...
+        &self,
+        dc_id: i32,
+    ) -> Result<Option<Arc<FileDownloader>>, InvocationError> {
+        return Ok({
+            let guard = self.0.downloader_map.read().await;
+            guard.get(&dc_id).map(|f| f.clone())
+        });
+    }
+
+    pub async fn invoke_in_dc<R: tl::RemoteCall>(
```
Comment (Lonami) [inline lib/grammers-client/src/client/net.rs:361]: I think it's fine to leave this as public (there's other requests which must be executed in a different datacenter, like getting graph data from channels I think).

File: lib/grammers-client/src/client/net.rs (hunk 6 above)
Comment (guzkodes) [reply lib/grammers-client/src/client/net.rs]: marked as resolved, so keep panic?

**Comment by Lonami** [comment]
Thanks! If you ever feel like you need to change something regarding this don't hesitate to send another pull request.
